src/databricks/labs/ucx/source_code/dependencies.py

```diff
@@ -1,77 +1,97 @@
 from __future__ import annotations
 
 import abc
+import ast
 from collections.abc import Callable, Iterable
+from pathlib import Path
 
-from databricks.sdk.service.workspace import ObjectType, ObjectInfo, ExportFormat, Language
+from databricks.sdk.service.workspace import ObjectType, ObjectInfo, ExportFormat
 from databricks.sdk import WorkspaceClient
 
 from databricks.labs.ucx.source_code.base import Advice, Deprecation
+from databricks.labs.ucx.source_code.python_linter import ASTLinter, PythonLinter
 from databricks.labs.ucx.source_code.whitelist import Whitelist, UCCompatibility
 
 
-class Dependency:
+class Dependency(abc.ABC):
 
-    @staticmethod
-    def from_object_info(object_info: ObjectInfo):
-        assert object_info.object_type is not None
-        assert object_info.path is not None
-        return Dependency(object_info.object_type, object_info.path)
-
-    def __init__(self, object_type: ObjectType | None, path: str):
-        self._type = object_type
+    def __init__(self, loader: DependencyLoader, path: Path):
+        self._loader = loader
         self._path = path
 
     @property
-    def type(self) -> ObjectType | None:
-        return self._type
-
-    @property
-    def path(self) -> str:
+    def path(self) -> Path:
         return self._path
 
     def __hash__(self):
         return hash(self.path)
 
     def __eq__(self, other):
-        return isinstance(other, Dependency) and self.path == other.path
+        return isinstance(other, type(self)) and self.path == other.path
+
+    def load(self) -> SourceContainer | None:
+        return self._loader.load_dependency(self)
 
 
 class SourceContainer(abc.ABC):
 
-    @property
     @abc.abstractmethod
-    def object_type(self) -> ObjectType:
+    def build_dependency_graph(self, parent: DependencyGraph) -> None:
         raise NotImplementedError()
 
+
+class DependencyLoader(abc.ABC):
+
     @abc.abstractmethod
-    def build_dependency_graph(self, graph) -> None:
+    def load_dependency(self, dependency: Dependency) -> SourceContainer | None:
         raise NotImplementedError()
 
+    @abc.abstractmethod
+    def is_notebook(self, path: Path) -> bool:
+        raise NotImplementedError()
+
+
+class LocalFileLoader(DependencyLoader):
+    def load_dependency(self, dependency: Dependency) -> SourceContainer | None:
+        raise NotImplementedError()
+
+    def is_file(self, path: Path) -> bool:
+        raise NotImplementedError()
+
+    def is_notebook(self, path: Path) -> bool:
+        raise NotImplementedError()
```

Comment:
```suggestion
```
file loader always loads files

Reply:
it's not loading a notebook, it's making sure it's not loading a notebook as a file by accident